### PR TITLE
Reland: Fix CUDA device guard usage when first arg of kernel is scalar

### LINKED ISSUE #38889: Normal.icdf is differs on different cuda devices
## 🐛 Bug

Normal.icdf is different on different cuda devices

[normal_cdf_bug.ipynb.txt](https://github.com/pytorch/pytorch/files/4664656/normal_cdf_bug.ipynb.txt)

```
import torch
x = torch.rand(10, dtype=torch.double)
normal = torch.distributions.Normal(0,1)
normal.icdf(x)
```
Output: 
>>> tensor([-0.0274,  0.9183,  0.8124,  0.1397, -0.1859,  0.0392, -0.3418,  0.8999,
         2.2092, -0.5433], dtype=torch.float64)

```
normal = torch.distributions.Normal(0,1)
normal.icdf(x.cuda())
```
Output:
>>> tensor([-0.0274,  0.9183,  0.8124,  0.1397, -0.1859,  0.0392, -0.3418,  0.8999,
     2.2092, -0.5433], dtype=torch.float64)
```
normal = torch.distributions.Normal(0,1)
normal.icdf(x.cuda(1))
```
Output:
>>> tensor([-0.0219,  0.6415,  0.5835,  0.1111, -0.1475,  0.0313, -0.2675,  0.6318,
         0.9728, -0.4131], device='cuda:1', dtype=torch.float64)
## Expected behavior

I expected Normal.icdf to be the same on all devices.

## Environment
Collecting environment information...
PyTorch version: 1.6.0a0-fb
Is debug build: No
CUDA used to build PyTorch: 9.2.88

OS: CentOS Linux 7 (Core)
GCC version: (GCC) 4.8.5 20150623 (Red Hat 4.8.5-39)
CMake version: Could not collect

Python version: 3.7
Is CUDA available: Yes
CUDA runtime version: Could not collect
GPU models and configuration: 
GPU 0: Tesla V100-SXM2-16GB
GPU 1: Tesla V100-SXM2-16GB
GPU 2: Tesla V100-SXM2-16GB
GPU 3: Tesla V100-SXM2-16GB
GPU 4: Tesla V100-SXM2-16GB
GPU 5: Tesla V100-SXM2-16GB
GPU 6: Tesla V100-SXM2-16GB
GPU 7: Tesla V100-SXM2-16GB

Nvidia driver version: 396.69
cuDNN version: Could not collect

Versions of relevant libraries:
[pip] Could not collect
[conda] Could not collect

cc @ezyang @gchanan @zou3519 @ngimel

### PR DESCRIPTION
**Pull request by kurtamohler**
Reland PR #39870 

Closes #38889

### CODE REVIEW

**Comment by kurtamohler** [comment]
Turns out that I'm getting test failures on `TestTorchDeviceTypeCUDA.test_serialization_cuda` only if I run all of `test_torch.py`. If I run just that failing test, it doesn't fail.

**Comment by ngimel** [comment]
Yeah, that happens because you are changing device in your test and don't set it back. You can either explicitly set the device back to origina, or, better, use `with torch.device` context manager.

**Comment by kurtamohler** [comment]
Oh right, that makes sense.

File: test/test_torch.py [hunk 1]
```diff
@@ -14166,6 +14166,24 @@ def test_cross_device_binary_ops(self, devices):
...
 
+    # This test ensures that a scalar Tensor can be safely used
+    # in a binary operation in conjuction with a Tensor on all
+    # available CUDA devices
+    @deviceCountAtLeast(2)
+    @onlyCUDA
+    def test_binary_op_scalar_device_unspecified(self, devices):
+        scalar_val = torch.tensor(1.)
+        for default_device in devices:
+            with torch.cuda.set_device(default_device):
```
Comment (ngimel) [inline test/test_torch.py:14177]: Does set_device work? I thought it is `torch.cuda.device(default_device)`.

Comment (kurtamohler) [reply test/test_torch.py:14177]: Yeah, it works, although the docs say "Usage of this function is discouraged in favor of `device`." Not sure why that is, but I'll change it.

Comment (kurtamohler) [reply test/test_torch.py]: Oh you're right, `cuda.set_device` doesn't create a context manager.

Comment (ngimel) [reply test/test_torch.py]: I'll wait for the tests, and should be good to merge.

**Review by ngimel — APPROVED** [review]

**Pull request closed without merging**